### DISCUSSION TITLE
AzureAD SSO - Groups

### DISCUSSION THREAD

**Discussion by umarkhan99** [netbox-community/netbox#12688] [Help Wanted!]
Hi,
I ahve configured AzureAD SSO for Netbox but don't know where/how to specify superuser/admin user groups. Please assist. I have netbox version 3.2.5 installed.

**Comment by umarkhan99** [comment]
Update: Added the following and still no success,

AUTH_AZUREAD_ENABLED = True                                                                                                            
REMOTE_AUTH_BACKEND = 'social_core.backends.azuread.AzureADOAuth2'                                                                     
SOCIAL_AUTH_AZUREAD_OAUTH2_KEY = 'APP ID'                                                                
SOCIAL_AUTH_AZUREAD_OAUTH2_SECRET = 'Secret Key'                                                         
AUTH_AZUREAD_REPLY_URL = 'https://HOSTNAME/login'                                                                       
AUTH_AZUREAD_ADMIN_GROUP = 'Netbox Admin'

**Comment by umarkhan99** [comment]
Correction: Added this part.

AUTH_AZUREAD_REPLY_URL = 'https://hostname/login'
AUTH_AZUREAD_ADMIN_GROUP = 'Netbox Admin'

**Comment by jcollie** [comment]
This is what I did to get it working. Although it's not well tested yet, and could probably be optimized some.

Add this to your config:

```python
REMOTE_AUTH_AUTO_CREATE_USER = True
REMOTE_AUTH_BACKEND = [
    "social_core.backends.azuread.AzureADOAuth2",
]
REMOTE_AUTH_DEFAULT_GROUPS = []
REMOTE_AUTH_DEFAULT_PERMISSIONS = {}
REMOTE_AUTH_ENABLED = True
SOCIAL_AUTH_AZUREAD_OAUTH2_KEY = "xxxx"
SOCIAL_AUTH_AZUREAD_OAUTH2_SECRET = "yyyy"
SOCIAL_AUTH_AZUREAD_OAUTH2_TENANT_ID = "zzzz"
SOCIAL_AUTH_AZUREAD_OAUTH2_DEFAULT_SCOPE = [
    "openid",
    "profile",
    "user_impersonation",
    "email",
    "groups",
]
SOCIAL_AUTH_AZUREAD_OAUTH2_EXTRA_DATA = [
    ("groups", "groups")
]
SOCIAL_AUTH_REDIRECT_IS_HTTPS = True
SOCIAL_AUTH_PIPELINE = (
    "social_core.pipeline.social_auth.social_details",
    "social_core.pipeline.social_auth.social_uid",
    "social_core.pipeline.social_auth.social_user",
    "social_core.pipeline.user.get_username",
    "social_core.pipeline.social_auth.associate_by_email",
    "social_core.pipeline.user.create_user",
    "social_core.pipeline.social_auth.associate_user",
    "netbox.authentication.user_default_groups_handler",
    "social_core.pipeline.social_auth.load_extra_data",
    "social_core.pipeline.user.user_details",
    "netbox.azuresetgroups.set_groups",
)
```

Copy this file to `/opt/netbox/netbox/netbox` (or wherever the NetBox code lives on your system). Probably better long-term adding this as a proper Python package but this works for now. You'll need to make sure that your Azure app registration sends the groups as the sAMAccountName or you'll need to use the Azure Group ID to map users to groups. You'll also need to customize this a bit for your specific group names.

```python
from django.contrib.auth.models import Group
import logging


class AuthFailed(Exception):
    pass


logger = logging.getLogger(__name__)


def set_groups(response, user, backend, *args, **kwargs):
    user.groups.clear()
    user.save()

    conndetails = user.social_auth.get(provider="azuread-oauth2")
    group_names = conndetails.extra_data.get("groups")
    logger.info("found groups: %s", group_names)

    if group_names is None:
        return

    try:
        user.is_active = True
        user.is_superuser = False
        user.is_staff = False
        user.save()

        for group_name in group_names:
            if group_name == "netbox superusers":
                user.is_superuser = True
                user.is_staff = True
                user.save()

            group, created = Group.objects.get_or_create(name=group_name)
            group.user_set.add(user)
            group.save()

    except Group.DoesNotExist:
        pass
```

**Comment by umarkhan99** [comment]
is this the only place where we specify the group name?

            if group_name == "netbox superusers":

**Reply by jcollie** [reply]
> is this the only place where we specify the group name?

Yes.

**Comment by umarkhan99** [comment]
Also getting this error message after making this change:

<class 'AttributeError'>

'list' object has no attribute 'rsplit'

Python version: 3.9.10
NetBox version: 3.2.5

**Reply by jcollie** [reply]
I'm using 3.5, so not sure what has changed. If you had a full backtrace that would be more useful.

**Comment by umarkhan99** [comment]
fixed the code with the following

REMOTE_AUTH_AUTO_CREATE_USER = True

REMOTE_AUTH_BACKEND = "social_core.backends.azuread.AzureADOAuth2"

REMOTE_AUTH_DEFAULT_GROUPS = []

REMOTE_AUTH_DEFAULT_PERMISSIONS = {}

REMOTE_AUTH_ENABLED = True

SOCIAL_AUTH_AZUREAD_OAUTH2_KEY = "xxxx"

SOCIAL_AUTH_AZUREAD_OAUTH2_SECRET = "yyyy"

SOCIAL_AUTH_AZUREAD_OAUTH2_TENANT_ID = "zzzz"

SOCIAL_AUTH_AZUREAD_OAUTH2_DEFAULT_SCOPE = [
    "openid",
    "profile",
    "user_impersonation",
    "email",
    "groups",
]
SOCIAL_AUTH_AZUREAD_OAUTH2_EXTRA_DATA = [
    ("groups", "groups")
]
SOCIAL_AUTH_REDIRECT_IS_HTTPS = True
SOCIAL_AUTH_PIPELINE = (
    "social_core.pipeline.social_auth.social_details",
    "social_core.pipeline.social_auth.social_uid",
    "social_core.pipeline.social_auth.social_user",
    "social_core.pipeline.user.get_username",
    "social_core.pipeline.social_auth.associate_by_email",
    "social_core.pipeline.user.create_user",
    "social_core.pipeline.social_auth.associate_user",
    "netbox.authentication.user_default_groups_handler",
    "social_core.pipeline.social_auth.load_extra_data",
    "social_core.pipeline.user.user_details",
    "netbox.azuresetgroups.set_groups",
)

**Reply by jcollie** [reply]
So you changed REMOTE_AUTH_BACKEND from a list to a single string? The formatting makes it difficult to see.

**Reply by umarkhan99** [reply]
ya, couldn't really find a way to fix the formatting so just added empty lines in the beginning

**Reply by jcollie** [reply]
Looks like REMOTE_AUTH_BACKEND behavior changed in 3.4.7: https://github.com/netbox-community/netbox/blob/b31b086a4d6518d143f2f1837391379ea9b94fc9/docs/release-notes/version-3.4.md?plain=1#L74

**Reply by umarkhan99** [reply]
that makes sense...do you know why I would receive the error message below?

**Comment by umarkhan99** [comment]
Getting this error hwoever:

AADSTS700016: Application with identifier 'xxxx' was not found in the directory 'DOMAIN.com'. This can happen if the application has not been installed by the administrator of the tenant or consented to by any user in the tenant. You may have sent your authentication request to the wrong tenant.

**Reply by jcollie** [reply]
"xxxx", "yyyy", and "zzzz" are placeholders. You'll need to replace them with the actual values from your Azure AD system.

**Reply by umarkhan99** [reply]
Forgot about that part lol....getting this error now:

<class 'AttributeError'>

module 'netbox.authentication' has no attribute 'user_default_groups_handler'

**Reply by jcollie** [reply]
You'll need to check the code to see what the default value for `SOCIAL_AUTH_PIPELINE` is for your version and add `netbox.azuresetgroups.set_groups` to the end. Or just remove `netbox.authentication.user_default_groups_handler` from `SOCIAL_AUTH_PIPELINE` and see what happens.

**Reply by umarkhan99** [reply]
I removed it and getting this error now:

A module import error occurred during this request. Common causes include the following:

Missing required packages - This installation of NetBox might be missing one or more required Python packages. These packages are listed in requirements.txt and local_requirements.txt, and are normally installed as part of the installation or upgrade process. To verify installed packages, run pip freeze from the console and compare the output to the list of required packages.

WSGI service not restarted after upgrade - If this installation has recently been upgraded, check that the WSGI service (e.g. gunicorn or uWSGI) has been restarted. This ensures that the new code is running.

The complete exception is provided below:

<class 'ModuleNotFoundError'>

No module named 'netbox.azuresetgroups'

**Reply by jcollie** [reply]
The python file needs to be named "azuresetgroups.py". Guess that I forgot to include that.

**Reply by umarkhan99** [reply]
at least I'm not getting any errors now ;). Still don't have admin privileges tho.

**Reply by umarkhan99** [reply]
where do I find the code that include netbox.authentication.user_default_groups_handler for v3.2.5?

**Reply by jcollie** [reply]
> where do I find the code that include netbox.authentication.user_default_groups_handler for v3.2.5?

You'll need to upgrade to a newer version of NetBox. I doubt that it would be very easy to backport that code to 3.2.5.

**Reply by jcollie** [reply]
> at least I'm not getting any errors now ;). Still don't have admin privileges tho.

The groups that Azure sends to NetBox should be logged. Check the logs for those and make sure that they make sense.

**Reply by umarkhan99** [reply]
where do I specify the groups in Azure that need to be imported? the documentation isn't really clear.

**Reply by umarkhan99** [reply]
I added them in Users and Groups for the Netbox application under Enterprise Application tab
